### PR TITLE
feat(formatters): move formatters to a separate package

### LINKED ISSUE #2467: Expose `formatters` from either a separate package or the CLI package
**User story.**
As a developer who wants to use Spectral in JS/TS, I can do create custom formatting, so that I can format my text output from an internal command without subshelling out to `spectral-cli`.

**Is your feature request related to a problem?**
I'm looking to use the existing formatters in `sepctral-cli` in Javascript but cannot as they are not exported from the `cli` package.

**Describe the solution you'd like**
Expose the formatters from their own package `@stoplight/spectral-formatters`. Alternatively, they could be exposed from the CLI, but would prefer their own package.

**Additional context**
I'm happy to submit a PR, this _should_ be pretty straightforward.

### PR DESCRIPTION
**Pull request by sennyeya**
Fixes #2467. Moves the formatters to their own package that is then imported by `@stoplight/spectral-cli`.

👀 @P0lip 

**Checklist**

- [ ] Tests added / updated
- [ ] Docs added / updated

Existing tests should cover any issues, this PR is just a move with additional dependency work.

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

The formatters for the CLI have been pulled out to their own package, not necessarily breaking for any contracts, but will require pulling in a new package on CLI version update.

**Additional context**

I wasn't sure if this should sit in `packages/core` or a separate package, I think it makes sense to pull it completely out as it's not core functionality and just a presentation layer, but I could go either way.

### CODE REVIEW

File: packages/formatters/CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# [@stoplight/spectral-parsers-v1.0.2](https://github.com/stoplightio/spectral/compare/@stoplight/spectral-parsers-v1.0.1...@stoplight/spectral-parsers-v1.0.2) (2022-08-22)
```
Comment (sennyeya) [inline packages/formatters/CHANGELOG.md]: Wasn't sure what this file should contain, I pulled the initial version from `packages/parsers`.

Comment (P0lip) [reply packages/formatters/CHANGELOG.md]: leave it empty for now

File: packages/formatters/README.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+const spectral = new Spectral();
+...
+const result = await spectral.run(document);
+
+// You can also filter the results here.
+const output = pretty(result);
+
+console.error(output);
+...
```
Comment (P0lip) [inline packages/formatters/README.md]:
```suggestion
// ...
```
maybe let's add a comment here to make sure the code run without any changes

File: packages/formatters/README.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+This project exposes the available formatters from the CLI for users that perform custom validation through Javascript.
+
+## Usage
+
+```ts
+import {pretty} from '@stoplight/spectral-formatters';
+
+const spectral = new Spectral();
+...
```
Comment (P0lip) [inline packages/formatters/README.md]:
```suggestion
// ...
```
maybe let's add a comment here to make sure the code run without any changes

File: package.json [hunk 4]
```diff
@@ -25,8 +25,7 @@
   "scripts": {
     "clean": "rimraf .cache packages/*/{dist,.cache}",
     "prebuild": "yarn workspace @stoplight/spectral-ruleset-migrator prebuild",
-    "build": "yarn prebuild && tsc --build ./tsconfig.build.json && yarn postbuild",
-    "postbuild": "yarn workspace @stoplight/spectral-cli postbuild",
```
Comment (P0lip) [inline package.json]: we need postbuild for formatters here

File: packages/formatters/CHANGELOG.md (hunk 1 above)
Comment (sennyeya) [reply packages/formatters/CHANGELOG.md]: Done.

File: packages/formatters/README.md (hunk 2 above)
Comment (sennyeya) [reply packages/formatters/README.md]: Done.

File: packages/formatters/README.md (hunk 3 above)
Comment (sennyeya) [reply packages/formatters/README.md]: Done.

File: package.json (hunk 4 above)
Comment (sennyeya) [reply package.json]: Done.

**Comment by sennyeya** [comment]
Couple of questions:
1. What is the commit message supposed to look like? Not sure I'm following from the error message.
2. Is there somewhere that the `memfs` is populated with the HTML files during Karma tests, getting a not found error locally and in the CI, but not sure where that would be solved?

**Comment by sennyeya** [comment]
@P0lip Any ideas on why the Karma tests are failing? I think it's related to `memfs` mocks, but it seems odd that moving the directory causes this issue in the same test case.

**Comment by P0lip** [comment]
Yeah, that is most likely due to the HTML formatter relying on `fs` and it not being available in Karma launcher we use (it's plain Chrome browser in headless mode, we use webpack to bundle things and have some basic fs mock).
I'll try to update HTML formatter to not depend on `fs` at all so that we don't need to worry about anything.

**Comment by P0lip** [comment]
#2474 - WIP

**Comment by sennyeya** [comment]
Great, thanks!

**Comment by P0lip** [comment]
@sennyeya could you merge the latest changes when you have a minute? 🙏 that should unblock test-browser 🤞

File: packages/formatters/package.json [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  "files": [
+    "/dist"
+  ],
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/stoplightio/spectral.git"
+  },
+  "scripts": {
+    "release": "semantic-release -e semantic-release-monorepo",
+    "postbuild": "copyfiles -u 1 \"./src/html/*.html\" \"./dist/\""
```
Comment (P0lip) [inline packages/formatters/package.json]: this no longer exists. I replaced it with prebuild, prelint, etc.

File: packages/formatters/package.json [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    "chalk": "4.1.2",
+    "cliui": "7.0.4",
+    "eol": "0.9.1",
+    "lodash": "^4.17.21",
+    "strip-ansi": "6.0",
+    "text-table": "^0.2.0",
+    "tslib": "^2.5.0"
+  },
+  "devDependencies": {
+    "copyfiles": "^2.4.1",
```
Comment (P0lip) [inline packages/formatters/package.json]:
```suggestion
```

File: packages/formatters/package.json (hunk 5 above)
Comment (P0lip) [inline packages/formatters/package.json]:
```suggestion
    "pretest": "yarn prebuild",
    "prelint": "yarn prebuild",
    "prebuild": "node scripts/bundle-html-templates.mjs",
```

Comment (P0lip) [reply packages/formatters/package.json]: you may also need to copy over that script I added https://github.com/stoplightio/spectral/pull/2474/files#diff-c889100b2ca9f8c4d4ec6b7a1cf8514839031ca73fa701bc5bc8d92f61e5ac8f

Comment (sennyeya) [reply packages/formatters/package.json]: Updated

File: packages/formatters/package.json (hunk 6 above)
Comment (sennyeya) [reply packages/formatters/package.json]: Updated.

File: packages/formatters/package.json (hunk 5 above)
Comment (sennyeya) [reply packages/formatters/package.json]: Updated.

**Comment by sennyeya** [comment]
@P0lip It looks like `cliui` is having issues with the Karma browser tests, which I think is expected as it's just a nodejs module. I think this was excluded originally by including `packages/ci/**` in the `karma.config.ts` `exclude` section. As you'll be updating this package to be all environment friendly, I think it makes sense to skip this for now.

Other than that, everything should be good minus the commit message (what's the format for those?)

**Comment by P0lip** [comment]
>  Other than that, everything should be good minus the commit message (what's the format for those?)

it has to be accepted by [commitlint](https://commitlint.js.org/#/) per its config here https://github.com/stoplightio/spectral/blob/develop/commitlint.config.js

**Comment by P0lip** [comment]
I'll take a final look on Monday, and if it's all good I will proceed and merge it.
Thanks a lot of your time and contribution. Appreciate it a lot.

**Review by P0lip — DISMISSED** [review]

**Review by P0lip — APPROVED** [review]

**Comment by sennyeya** [comment]
@P0lip Thanks for your help pushing this through!

**Pull request merged**